EXTENSIONS.md

```diff
@@ -0,0 +1,236 @@
+
+## Annotation syntax
+
+**Annotations**, a declarative complement to decorators, use the syntax `@{ }`, which behaves exactly like an object literal. Arbitrary expressions, spread, computed property names, etc, are permitted. It is accessible from the annotated object through the `[Symbol.metadata]` property.
+
+```mjs
+@{x: "y"} @{v: "w"} class C {
+  @{a: "b"} method() { }
+  @{c: "d"} field;
+}
+
+C[Symbol.metadata].class.x                     // "y"
+C[Symbol.metadata].class.v                     // "w"
+C[Symbol.metadata].prototype.methods.method.a  // "a"
+C[Symbol.metadata].instance.fields.field.c     // "d"
+```
+
+Annotations must always have a literal `@{` to start them. To use an existing object as an annotation, you can use the syntax `@{ ...obj }`. The entirety of object syntax is available, including computed property names, arbitrary expressions as values, shorthand names, concise methods, etc.
+
+Libraries and frameworks which want to establish consistent conventions for using annotations may do so based on a Symbol property key that they export. Annotations have the potential advantage in load time performance that engines can directly execute them, as they are as declarative as an object literal.
+
+Annotation semantics may be useful for cases like ORMs and serialization frameworks, which need information about class fields, without affecting their normal runtime semantics. However, the popular ecosystem examples that we've found of this form, using just metadata for fields, seem to depend on metadata generated by TypeScript types. From these examples, it seems that annotation syntax alone would not be a sufficient solution.
+
+Some frameworks, including Angular, tend to use decorators which operate primarily by adding metadata. However, object literal annotations are not quite suitable for this usage, as they don't provide a way to be annotated in TypeScript to check types the way that functions do, they don't allow any processing of the metadata in code before it is saved, and because they don't provide a usable, stable identifier to be used for custom static analysis tools like tree shaking. For Angular, it may make more sense to use decorators which add metadata.
+
+For these reasons, annotations are omitted from this proposal's "MVP" (minimum viable product) and considered as a possible follow-on proposal.
+
+## Function decorators and annotations
+
+The `@logged` decorator from earlier would Just Work(TM) on a function, with function decorators!
+
+```js
+@logged
+@{x: "y"}
+function f() { }
```

Comment:
I figure I will explain my tweet since it may not have occurred to folks&hellip;

> What I mean to say is that this seems to be proposed with complete disregard for JSDoc syntax.
> &mdash; https://twitter.com/DerekNonGeneric/status/1303036574893047809

---

Obviously, I'm not sold on the *decorator* idea, but it's not the concept that I am in disagreement with, it's the distasteful meshing with JSDoc that I foresee. How are people are okay with putting decorators above function and class declarations like this? That's where JSDoc comments go. None of these examples take JSDoc into account, do they? My intention wasn't to be snarky and I don't think the concern I raised is misguided. Can someone perhaps clarify how this isn't going to make for exceptionally unsightly codebases? Sometimes I feel like I'm the only person representing the JSDoc userbase while everyone is trying to make it obsolete/irrelevant.

Reply:
I think you are off base here. The `@` symbol was deliberated and settled on several years ago. It won't conflict with jsdoc since doc tags are in comment bodies.

Reply:
Maybe I was unclear, perhaps the following will help.

> In JavaScript files, TypeScript will only consult immediately preceding JSDoc comments to figure out declared types.
> &mdash; https://devblogs.microsoft.com/typescript/announcing-typescript-3-6/#jsdoc-comments-no-longer-merge

Would it be possible to include an example with a JSDoc comment above a decorated class and/or function?

Reply:
Do you mean something like this?
```js
/** 
 * JSDoc comment...
 * @see {@link http://github.com|GitHub}
 */
@registerElement('something') 
class Example extends HTMLElement {

  /**
   * @param {string} target
   */
  @someTrait doSomething = (target) => {}

}
```

The decorator is part of the class/property definition, so the JSDock is still immediately preceding it.

Reply:
Functions are not part of this proposal. But, it would be no different, the decorator list would be a part of the definition so the comment block can still be immediately preceding the definition

Reply:
Yeah, something like this would be helpful. Thanks!